File: DSA/VaniJain_590012020/Day18Question3.java
import java.util.*;
public class DutchNationalFlag 
{
    public static void main(String[] args) 
    {
        Scanner scanner = new Scanner(System.in);
        System.out.print("Enter the number of elements: ");
        int n = scanner.nextInt();
        int[] arr = new int[n];
        System.out.println("Enter the elements (only 0, 1, or 2):");
        for (int i = 0; i < n; i++) 
        {
            arr[i] = scanner.nextInt();
        }
        int low = 0, mid = 0, high = n - 1;

        while (mid <= high) 
        {
            if (arr[mid] == 0) 
            {
                int temp = arr[low];
                arr[low] = arr[mid];
                arr[mid] = temp;
                low++;
                mid++;
            } else if (arr[mid] == 1)
            {
                mid++; 
            } else 
            { 
                int temp = arr[mid];
                arr[mid] = arr[high];
                arr[high] = temp;
                high--;
            }
        }
        System.out.println("Sorted array:");
        for (int i = 0; i < n; i++) 
        {
            System.out.print(arr[i] + " ");
        }
    }
}

File: DSA/Prabhjeet_590013286/Day18/Day_18_Question_3.c
// 3: Sort Colors (Dutch National Flag)
// Given an array with elements 0, 1, 2, sort them efficiently in one pass.
// Place all 0s first, followed by 1s, then 2s.

#include <stdio.h>

void func(int arr[], int n) {
    int low = 0, mid = 0, high = n - 1;
    while (mid <= high) {
        if (arr[mid] == 0) {
            int temp = arr[low];
            arr[low++] = arr[mid];
            arr[mid++] = temp;
        } else if (arr[mid] == 1) {
            mid++;
        } else {
            int temp = arr[mid];
            arr[mid] = arr[high];
            arr[high--] = temp;
        }
    }
}

int main() {
    int n;
    printf("Enter number of elements: ");
    scanf("%d", &n);
    int arr[n];
    printf("Enter elements (0, 1, 2 only):\n");
    for (int i = 0; i < n; i++)
        scanf("%d", &arr[i]);

    func(arr, n);

    printf("Sorted array: ");
    for (int i = 0; i < n; i++)
        printf("%d ", arr[i]);
    return 0;
}